src/System.Linq/src/System/Linq/Enumerable.SelectWithCountIterator.cs

```diff
@@ -0,0 +1,62 @@
+﻿
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract class SelectWithCountIterator<TSource> : SelectIterator<TSource>, ICollection<TSource>
+        {
+            public override abstract Iterator<TSource> Clone();
+
+            public override abstract bool MoveNext();
+
+            public override abstract IEnumerable<TSource> Where(Func<TSource, bool> predicate);
+
+            public override abstract IEnumerable<TResult> SelectImpl<TResult>(Func<TSource, TResult> selector);
+
+            #region ICollection<TSource> implementation
+
+            public abstract int Count { get; }
+
+            public bool IsReadOnly { get { return true; } }
+
+            public void Add(TSource item)
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public void Clear()
+            {
+                throw NotImplemented.ByDesign;
+            }
+
+            public bool Contains(TSource item)
+            {
+                throw NotImplemented.ByDesign;
```

Comment:
It would be great to see the actual gain from ```CopyTo```. Can you move its implementation to ```SelectArrayIterator``` and to ```SelectListIterator``` and measure the performance again? According to my expectations, we will get extra 2x speed-up. If the gain will be too small we can just ignore ```CopyTo```.